File: java-to-go-projects/13-event-emitter/Emitter.java
package events;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.function.Consumer;

/**
 * A thread-safe event emitter supporting:
 *   - on(event, callback)           — Add a persistent listener
 *   - once(event, callback)         — Add a one-shot listener
 *   - off(event, listenerId)        — Remove a listener by ID
 *   - emit(event, data)             — Fire event to all listeners
 *   - listenerCount(event)          — Count listeners for an event
 *   - eventNames()                  — List all registered event names
 *
 * Supports wildcard event "*" that receives all events.
 */
public class Emitter<T> {

    private final Map<String, List<Listener<T>>> listeners = new ConcurrentHashMap<>();
    private int nextId = 0;

    /**
     * Register a persistent listener for an event.
     * Returns the listener ID for later removal.
     */
    public String on(String event, Consumer<T> callback) {
        return addListener(event, callback, false);
    }

    /**
     * Register a one-shot listener (auto-removed after first emit).
     */
    public String once(String event, Consumer<T> callback) {
        return addListener(event, callback, true);
    }

    /**
     * Remove a specific listener by ID from an event.
     */
    public boolean off(String event, String listenerId) {
        List<Listener<T>> list = listeners.get(event);
        if (list == null) return false;
        return list.removeIf(l -> l.getId().equals(listenerId));
    }

    /**
     * Remove ALL listeners for an event.
     */
    public void offAll(String event) {
        listeners.remove(event);
    }

    /**
     * Emit an event with associated data. Invokes all matching listeners.
     * Also invokes wildcard "*" listeners.
     */
    public void emit(String event, T data) {
        fire(event, data);
        if (!"*".equals(event)) {
            fire("*", data); // wildcard listeners
        }
    }

    /**
     * Number of listeners registered for a specific event.
     */
    public int listenerCount(String event) {
        List<Listener<T>> list = listeners.get(event);
        return list == null ? 0 : list.size();
    }

    /**
     * All event names that have at least one listener.
     */
    public Set<String> eventNames() {
        return listeners.keySet();
    }

    private void fire(String event, T data) {
        List<Listener<T>> list = listeners.get(event);
        if (list == null) return;

        List<Listener<T>> toRemove = new ArrayList<>();
        for (Listener<T> listener : list) {
            boolean shouldRemove = listener.invoke(data);
            if (shouldRemove) {
                toRemove.add(listener);
            }
        }
        list.removeAll(toRemove);

        if (list.isEmpty()) {
            listeners.remove(event);
        }
    }

    private String addListener(String event, Consumer<T> callback, boolean once) {
        String id = "listener_" + (nextId++);
        listeners.computeIfAbsent(event, k -> new CopyOnWriteArrayList<>())
                 .add(new Listener<>(id, callback, once));
        return id;
    }
}

File: java-to-go-projects/13-event-emitter/rust/src/emitter.rs
use std::collections::HashMap;
use crate::listener::Listener;

pub struct Emitter<T> {
    listeners: HashMap<String, Vec<Listener<T>>>,
    next_id: usize,
}

impl<T> Emitter<T> {
    pub fn new() -> Self {
        todo!()
    }

    /// Register a persistent listener. Returns listener id.
    pub fn on(&mut self, event: &str, callback: Box<dyn Fn(&T) + Send + Sync>) -> usize {
        todo!()
    }

    /// Register a one-shot listener. Returns listener id.
    pub fn once(&mut self, event: &str, callback: Box<dyn Fn(&T) + Send + Sync>) -> usize {
        todo!()
    }

    /// Remove a specific listener by event name and id. Returns true if found.
    pub fn off(&mut self, event: &str, id: usize) -> bool {
        todo!()
    }

    /// Remove all listeners for an event.
    pub fn off_all(&mut self, event: &str) {
        todo!()
    }

    /// Emit an event. Invokes all matching listeners + wildcard "*" listeners.
    /// Removes one-shot listeners after invocation.
    pub fn emit(&mut self, event: &str, data: &T) {
        todo!()
    }

    /// Count listeners for a specific event.
    pub fn listener_count(&self, event: &str) -> usize {
        todo!()
    }

    /// Return all event names that have listeners.
    pub fn event_names(&self) -> Vec<String> {
        todo!()
    }
}
